packages/core/src/tracing/instrumentation/database/db2.js

```diff
@@ -0,0 +1,612 @@
+
+  // CASE: There can be only one exit span, skip
+  if (constants.isExitSpan(parentSpan)) {
+    return originalFunction.apply(ctx, originalArgs);
+  }
+
+  return cls.ns.runAndReturn(() => {
+    const span = createSpan(stmt, instrumentQueryResultHelper);
+
+    if (!isAsync) {
+      try {
+        const result = originalFunction.apply(ctx, originalArgs);
+        finishSpan(ctx, result, span);
+        return result;
+      } catch (err) {
+        span.ec = 1;
+        span.data.db2.error = tracingUtil.getErrorDetails(err);
+        finishSpan(ctx, null, span);
+        return err;
+      }
+    }
+
+    const customerCallbackIndex =
+      // eslint-disable-next-line no-nested-ternary
+      originalArgs.length === 2 && typeof originalArgs[1] === 'function'
+        ? 1
+        : originalArgs.length === 3 && typeof originalArgs[2] === 'function'
+        ? 2
+        : null;
+    const customerCallback = originalArgs[customerCallbackIndex];
+
+    if (customerCallback) {
+      originalArgs[customerCallbackIndex] = function instanaCallback(err) {
+        if (err) {
+          span.ec = 1;
+          span.data.db2.error = tracingUtil.getErrorDetails(err);
+        }
+
+        const result = customerCallback.apply(this, arguments);
+        finishSpan(ctx, result, span);
+        return result;
+      };
+
+      return originalFunction.apply(ctx, originalArgs);
+    }
+  });
+}
+
+function createSpan(stmt, fn) {
+  // eslint-disable-next-line max-len
+  // https://github.ibm.com/instana/backend/blob/develop/forge/src/main/java/com/instana/forge/connection/database/ibmdb2/IbmDb2Span.java
+  const span = cls.startSpan(exports.spanName, constants.EXIT);
+  span.stack = tracingUtil.getStackTrace(fn);
+  span.d = Date.now() - span.ts;
+  span.data.db2 = {
+    stmt: tracingUtil.shortenDatabaseStatement(stmt),
+    dsn: tracingUtil.sanitizeConnectionStr(connectionStr)
+  };
+
+  return span;
+}
+
+function finishSpan(ctx, result, span) {
+  captureFetchError(result, span);
+
+  if (ctx.conn.inTransaction) {
+    return handleTransaction(ctx, span);
+  }
+
+  // NOTE: This signalises us the end of the trace
+  //       Because we want to capture errors till the customer finished the operation.
+  //       That's why we want to wait for this call. The ibm docs say, it's required to call it!
+  if (result && result.closeSync) {
+    const originalCloseSync = result.closeSync;
+    let closeSyncCalled = false;
+    result.closeSync = function instanaCloseSync() {
+      if (closeSyncCalled) return originalCloseSync.apply(this, arguments);
+
+      closeSyncCalled = true;
+      span.transmit();
+      return originalCloseSync.apply(this, arguments);
+    };
+
+    // CASE: customer forgets to call `result.closeSync`
+    //       search for result.closeSync() in:
+    //       https://github.com/ibmdb/node-ibm_db/blob/master/APIDocumentation.md
+    // TODO: discuss timeout length and the approach in general
+    setTimeout(() => {
```

Comment:
Discussed:
`clearTimeout(t)` is missing when timeout was called
`unref` is missing e.g. when app is killed and should not block the runtime